### PR TITLE
better default inits

### PR DESCRIPTION
**Pull request by kevinzakka**
* linear layers: `xavier_uniform` for weights, `zeros` for biases
* conv layers: `xavier_uniform` for weights, `zeros` for biases
* recurrent layers: `xavier_uniform` for ih weight, `orthogonal` for hh weight, `zeros` for biases

### CODE REVIEW

**Comment by soumith** [comment]
i'm not sure we should merge this. what's the motivation?

Separately, I think we should make it easier to use `nn.init` in Linear, Conv and RNN layers by specifying strings in the constructors.

For example: `nn.Linear(20, 30, weight_init='xavier')`

**Comment by vishwakftw** [comment]
I think BatchNorm implicitly tries to make random initialization just as good as Xavier init. +1 for the idea on adding it the constructor.

**Comment by kevinzakka** [comment]
@soumith so I've fleshed out a PR where I've changed the `init.py` file to contain classes for each initialization and have added support for 2 string args in the layers: `weight_initializer` and `bias_initializer`. 

However, I had a discussion with @apaszke and he thought we should leave that to higher level libraries and keep things the way they are.

Should I submit a PR for it in case you guys want to see it?

**Comment by soumith** [comment]
@apaszke i see quite a lot of value in having `weight_init` directly in constructor of Conv / Linear / RNN. Can you expand your reservations?

**Comment by kevinzakka** [comment]
@soumith should I submit a separate PR and reference this one?

**Comment by elanmart** [comment]
@kevinzakka could you also add this to `nn.Embedding`? IIRC it had a pretty poor default scheme.

**Comment by apaszke** [comment]
@soumith I'm just saying that it's often not quite as easy as dropping in a string, as some functions might take extra parameters (e.g. the following non-linearity, bounds of the uniform distribution). I can't see a huge usability difference between adding it to the constructor, or adding
```python
nn.init.uniform(a=-0.5, b=-0.5)
```
the line below (with this approach having an extra benefit that you have more control over init).

**Comment by kevinzakka** [comment]
@apaszke in #5382 I've added the following options to the user:

- use a string to initialize weights and biases. This can't be parametrized but you would use this for initializations like `orthogonal`, `zeros`, `ones`, `kaiming_uniform` (assuming ReLU activation) and `xavier_uniform` (assuming a general-purpose good init for any activation).
- use a function provided in `init.py` which you can parametrize directly to your liking. This would be used for random inits like `uniform` and `normal`, and for `xavier_uniform` which you want to explicitly tailor to your activation using the `calculate_gain()` method.

```
x = nn.Linear(784, 512, weight_init='xavier_uniform', bias_init='zeros')
x = nn.Linear(784, 512, weight_init=init.uniform(-5, 5))
x = nn.Linear(784, 512, weight_init=init.xavier_uniform(gain=init.calculate_gain('tanh')))
```

**Comment by apaszke** [comment]
The problem is that this approach is incompatible with the signature of `nn.init` methods. For example, here's the definition of `uniform`:
```python
def uniform(tensor, a=0, b=1):
```
Your approach implies that doing `nn.uniform(-5, 5)` would return a function that would await a tensor argument, and fill it according to the policy, but this is not the case. The only reasonable short alternative would be to use `partial(init.uniform, a=-5, b=5)`, but that's a just ugly...

**Comment by kevinzakka** [comment]
@apaszke agreed, which is why I refactored the whole `init.py` file to contain classes which contain a `__call__` method on a given tensor.

**Comment by apaszke** [comment]
NB: if we were to allow both strings and functions, and strings would just do `getattr(nn.init, name)`, then I think we should force users to supply a function (so instead of `some_init='xavier'`, you'd do `some_init=init.xavier`). The benefits are quite obvious (think enums vs string values).

**Comment by apaszke** [comment]
@kevinzakka the problem with this approach is that it's not backwards compatible. We could support both the functional way, and the curried way be implementing a class with `__new__`. It would work like this:
- In `__new__`, check if you got the tensor. If yes, this is a old-style version, and we should just run the function (can be implemented as `cls(**non_tensor_args)(tensor_arg)`.
- Otherwise call`__init__` and save the specified arguments
- In `__call__` accept only a single tensor, and apply the arguments cached in `__init__`
